README.md

<img src='https://styled-system.com/logo.svg' width='128' height='128' />

# Styled System

Responsive, theme-based style props for building design systems with React
https://styled-system.com

[![stars][]][github]
[![Build Status][ci]][travis]
[![Downloads][]][npm]
[![Version][]][npm]
[![spectrum-badge][]][spectrum]
![size][]
![MIT License][license]
[![system-ui/theme][system-ui-badge]](https://system-ui.com/theme)

[ci]: https://flat.badgen.net/travis/styled-system/styled-system
[version]: https://flat.badgen.net/npm/v/styled-system
[downloads]: https://flat.badgen.net/npm/dm/styled-system
[license]: https://flat.badgen.net/badge/license/MIT/blue
[stars]: https://flat.badgen.net/github/stars/styled-system/styled-system
[size]: https://flat.badgen.net/bundlephobia/minzip/styled-system
[spectrum-badge]: https://flat.badgen.net/badge/spectrum/community/purple
[system-ui-badge]: https://flat.badgen.net/badge/system-ui/theme/black

[npm]: https://npmjs.com/package/styled-system
[travis]: https://travis-ci.org/styled-system/styled-system
[github]: https://github.com/styled-system/styled-system
[spectrum]: https://spectrum.chat/styled-system


```sh
npm i styled-system
```

## Features

- Add style props that hook into your own theme
- Quickly set responsive font-size, margin, padding, width, and more with props
- Influenced by constraint-based design system principles
- Typographic scale
- Spacing scale for margin and padding
- Works with any color palette
- Works with most css-in-js libraries, including [styled-components][] & [emotion][emotion]
- Used in [Rebass](https://rebassjs.org), [Rebass Grid](https://rebassjs.org/grid/), and the [Priceline Design System](https://github.com/pricelinelabs/design-system)

> "This is honestly my favourite way to build UI components right now ![party parrot][party-parrot]"
>
> – [Varun Vachhar][varun-post]

[party-parrot]: https://github.com/jmhobbs/cultofthepartyparrot.com/raw/master/parrots/parrot.gif


> "If you haven't seen Styled System before, do yourself a favour and check it out. It's been a huge influence in my thinking on component-oriented styles."
>
> – [Mark Dalgleish][markdalgleish]

<!--
> "Fantastic set of tools that offer the ease and API of tachyons/functional CSS but, are way more customisable."
>
> – [Varun Vachhar](https://mobile.twitter.com/winkerVSbecks/status/955619873463431168)
-->

> "The future of css-in-js is going to look something like styled-system with its responsive values."<br/>
>
> – [Kye Hohenberger](https://mobile.twitter.com/kyehohenberger/status/905474043729416192)

> "Coming from @tachyons_css, the styled-system utilities from @jxnblk is the missing link I’ve been looking for."<br/>
>
> – [Nathan Young](https://mobile.twitter.com/nathanyoung/status/891353221880360960)

> "If you make websites/apps with React check out Styled System if you haven't already. You will be amazed at how much faster you can build."
>
> – [David Yeiser][david-tweet]

> "If you like Tachyons you will love styled-system. If you don't like Tachyons, you will love styled-system."
>
> – [Adam Morse][mrmrs]

[mrmrs]: https://mobile.twitter.com/mrmrs_/status/913189805055401984


## Try It Out

Try the [examples](examples) on CodeSandbox

- [Basic Example](https://codesandbox.io/s/github/jxnblk/styled-system/tree/master/examples/basic)
- [Emotion](https://codesandbox.io/s/github/jxnblk/styled-system/tree/master/examples/emotion)

### Table of Contents

- [Usage](#usage)
- [Docs](#docs)
- [Further Reading](#further-reading)
- [Built with Styled System](#built-with-styled-system)
- [Related](#related)

## Usage

```jsx
// Example uses styled-components, but styled-system works with most other css-in-js libraries as well
import styled from 'styled-components'
import { space, width, fontSize, color } from 'styled-system'

// Add styled-system functions to your component
const Box = styled.div`
  ${space}
  ${width}
  ${fontSize}
  ${color}
`
```

Each style function exposes its own set of component props
that handle styles based on values defined in a theme.

```jsx
// width: 50%
<Box width={1/2} />

// font-size: 20px (theme.fontSizes[4])
<Box fontSize={4} />

// margin: 16px (theme.space[2])
<Box m={2} />

// padding: 32px (theme.space[3])
<Box p={3} />

// color
<Box color='tomato' />

// color: #333 (theme.colors.gray[0])
<Box color='gray.0' />

// background color
<Box bg='tomato' />
```

## Responsive Style Props

Set responsive width, margin, padding, font-size, and other properties with a shorthand array syntax.
[Read more](https://styled-system.com/responsive-styles)

```jsx
// responsive width
<Box width={[ 1, 1/2, 1/4 ]} />

// responsive font-size
<Box fontSize={[ 2, 3, 4 ]} />

// responsive margin
<Box m={[ 1, 2, 3 ]} />

// responsive padding
<Box p={[ 1, 2, 3 ]} />
```

To learn more, see the [Getting Started](https://styled-system.com/getting-started) guide or read the docs.

<!-- link shims for previous readme -->
<a name='getting-started'></a>
<a name='how-it-works'></a>
<a name='responsive-styles'></a>
<a name='api'></a>
<a name='system-components'></a>
<a name='default-theme'></a>
<a name='troubleshooting'></a>
<a name='cleanelement'></a>

## Docs

- [Getting Started](https://styled-system.com/getting-started)
- [Responsive Styles](https://styled-system.com/responsive-styles)
- [How it Works](https://styled-system.com/how-it-works)
- [API](https://styled-system.com/api)
- [Table of Style Functions](https://styled-system.com/table)
- [Custom Props](https://styled-system.com/custom-props)

---

## Further Reading

- [Component Based Design System With Styled-System][varun-post]
- [Build Better Component Libraries with Styled System][alan-b-smith-post]
- [Building a beautiful, reusable button with Styled System](https://johno.com/styled-system-button)
- [Defining Component APIs in React](http://jxnblk.com/blog/defining-component-apis-in-react/)
- [The Three Tenets of Styled System](https://jxnblk.com/blog/the-three-tenets-of-styled-system/)


## Built with Styled System

- [Primer Components](https://primer.style/components)
- [Priceline Design System](https://pricelinelabs.github.io/design-system/)
- [Cloudflare Design Color](https://cloudflare.design/color/)
- [Styled System HTML](https://johnpolacek.github.io/styled-system-html/)
- [Rebass](https://rebassjs.org)
- [Rebass Grid](https://grid.rebassjs.org)

## Related

- [System UI](https://system-ui.com)
- [Spectrum.Chat Community](https://spectrum.chat/styled-system)
- [styled-components][]
- [emotion][]

[styled-components]: https://github.com/styled-components/styled-components
[emotion]: https://github.com/emotion-js/emotion
[varun-post]: https://varun.ca/styled-system/
[alan-b-smith-post]: https://medium.com/styled-components/build-better-component-libraries-with-styled-system-4951653d54ee
[david-tweet]: https://mobile.twitter.com/davidyeiser/status/965920740582285312

<!-- extra links -->
[cole-tweet]: https://mobile.twitter.com/colebemis/status/996565848138526721
[mrmrs-elements]: https://github.com/mrmrs/elements
[broccs-react-starter]: https://github.com/broccolini/react-website-starter
[markdalgleish]: https://mobile.twitter.com/markdalgleish/status/1107732365474848768
[Varun Vachhar]: https://mobile.twitter.com/winkerVSbecks/status/955619873463431168

[MIT License](LICENSE.md)


CHANGELOG.md
# Changelog

## Unreleased

## v5.1.5 2020-02-17

- `@styled-system/css` fix media query order #1137
- Update dependencies

## v5.1.4 2019-12-02

- `@styled-system/css` add `outlineColor` support #825
- Update dependencies

## v5.1.3 2019-09-16

- Add `@styled-system/props` utility #824

## v5.1.2 2019-09-11

- Update dependencies
- Fix sort for media queries in plain object syntax #812

## v5.1.1 2019-08-22

- Fix for passing props through to transform function #723
- Update dependencies

## v5.1.0 2019-08-12

- New `variant` API with support for inline variant definitions and theme-based style objects #699

## v5.0.23 2019-08-12

- Add support for `overflowX` and `overflowY` #682
- `@styled-system/css`: add support for `size` #684
- Update dependencies

## v5.0.22 2019-08-09

- `@styled-system/css`: add support for `fill`, `stroke`, and `flexBasis` #679

## v5.0.21 2019-08-07

- Fix publish

## v5.0.20 2019-08-07

- Add theme support for more border properties #670 & #671

## v5.0.19 2019-08-05

- `@styled-system/css` allow functional values to return responsive array values #654

## v5.0.18 2019-07-31

- Fix for incorrectly ordered `@media` rules when `null` values are used in multiple array-value props #638
- Fix for incorrectly ordered `@media` rules with responsive object values
- Added integration tests with Styled Components and Emotion
- `@styled-system/prop-types` exports `createPropTypes`

## v5.0.17 2019-07-30

- Failed lerna publish

## v5.0.16 2019-07-22

- Update dependencies
- `@styled-system/css`: fix for `p` paragraph/padding collision
- Add option to disable breakpoint caching #595

## v5.0.15 2019-07-15

- Export single-property functions for v4 shims #584

## v5.0.14 2019-07-15

- Use `theme.space` for position props #598

## v5.0.13 2019-07-05

- `@styled-system/css`: add support for `gap`, `rowGap`, and `columnGap` properties #596

## v5.0.12 2019-06-25

- Fix for parser config

## v5.0.11 2019-06-24

- `@styled-system/css`: use `theme.space` in grid properties #574

## v5.0.10 2019-06-19

- `@styled-system/css`: use `theme.space` in position properties #568
- `@styled-system/css`: fix null check #572

## v5.0.9 2019-06-18

- `@styled-system/css`: use theme in `textShadow` property #567

## v5.0.8 2019-06-12

- Add missing prop types #557

## v5.0.7 2019-06-11

- `@styled-system/should-forward-prop`: add missing props #554
- Fix example #548

## v5.0.6 2019-06-10

- Update dependencies #553
- Add unit test #545
- Update docs #542, #543

## v5.0.5 2019-06-03

- Fix `get` utility #536
- Fix `boxShadow` export #535
- Fix typo in docs #534
- Fix typo in docs #530

## v5.0.4 2019-06-03

- Fix export #528
- Update docs #525, #526, #527

## v5.0.3 2019-06-03

- Fix margin transform

## v5.0.2 2019-06-02

- Add null check for `get` utility

## v5.0.1 2019-06-02

- Edit Babel config

## v5.0.0 2019-06-02

- Faster (fastest implementation yet)
- New `system` API for custom style props
- Built-in style props are now grouped into categories
- Style functions are now available as separate packages (e.g. `@styled-system/core`, `@styled-system/space`)
- Removes `propTypes` (use `.propNames` or the optional `@styled-system/prop-types` package instead)
- No longer converts numbers to `px` strings
- Changes interal `get` utility
- Removes: `isObject`, `is`, `px`, `num`, `createMediaQuery`, `defaultBreakpoints`, `merge`, `cloneFunction`, `mapProps`
- No longer returns `null` (returns and empty object instead)
- No longer converts padding to negative values (not needed)
- Removes `themeGet` from main package, use `@styled-system/theme-get` instead
- The new `theme.sizes` scale replaces `heights`, `minHeights`, `maxHeights`, `minWidths`, and `maxWidths`; and is used for the `width` prop
- Style functions now cache `theme.breakpoints`. If you rely on dynamically changing breakpoints in runtime, this is a breaking change

## v4.2.3 2019-05-27

- Fix for sort order in IE #465

## v4.2.2 2019-05-08

- Fix for `merge` utility #483

## v4.2.1 2019-05-04

- Fix build

## v4.2.0 2019-05-04

- Merge returned styles into a single object #473

## v4.1.0 2019-03-28

- #440
- #442
- #439
- #438

## v4.0.8 2019-03-23

- Fix `.npmignore`

## v4.0.7 2019-03-23

- #434
- #435
- #433
- #431

## v4.0.6 2019-03-21

- Changes source order in `borders` #429

## v4.0.5 2019-03-15

- Fix for missing `propTypes` on `space` function

## v4.0.1 2019-02-27

- Fix for props in `textStyles` and `colorStyles`

## v4.0.0 2019-02-23

- Rewritten core for less code duplication
- No longer merges returned style object, but returns arrays instead, which should work with common CSS-in-JS libraries
- `space` function has been rewritten with the core `style` and `compose` utilities
- Adds long-form props for margin and padding (e.g. `marginTop`). Shorthand aliases still work as before.
- Reintroduces the `alias` argument to `style` utility
- Removes the `styles` object export
- Removes the `util` object export
- Removes `merge` utility
- Removes the `mixed` utility since it did not behave as expected
- The theme "scale" is now passed as the second argument to the `transformValue` option in `style` utility
- Removes the `ratio` style function
- The following changes to the `borders` style function
  - The shorthand border props no longer transform numbers to pixel widths with a `solid` `border-style`
  - Adds `borderWidth`, `borderStyle`, `borderColor`, and `borderRadius` props
  - All border props are also available as individual imports
- Removes `styleType` from prop types `meta` fields
- Changes to the `get` utility: now returns the last argument as a fallback
- Rename `bgColor` to `backgroundColor`

## v3.2.1 2019-01-02

- Adds prop types for objects #366

## v3.2.0 2019-01-02

- Allow object syntax for responsive values #341

## v3.1.12 2019-01-02

- Allow nested objects for `theme.space` #352

## v3.1.11 2018-10-11

- Bugfix: remove usage of `in` operator in get utility

## v3.1.10 2018-10-10


src/v3.js
import PropTypes from 'prop-types'

// utils
const noop = n => n

export const propTypes = {
  numberOrString: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  responsive: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
}

export const defaultBreakpoints = [40, 52, 64].map(n => n + 'em')
export const is = n => n !== undefined && n !== null
export const num = n => typeof n === 'number' && !isNaN(n)
export const px = n => (num(n) ? n + 'px' : n)
export const isObject = n => typeof n === 'object' && n !== null

export const get = (obj, ...paths) =>
  paths
    .join('.')
    .split('.')
    .reduce((a, b) => (a && a[b] ? a[b] : null), obj)

export const themeGet = (paths, fallback) => props =>
  get(props.theme, paths) || fallback

export const cloneFunc = fn => (...args) => fn(...args)

export const merge = (a, b) =>
  Object.assign(
    {},
    a,
    b,
    Object.keys(b || {}).reduce(
      (obj, key) =>
        Object.assign(obj, {
          [key]:
            a[key] !== null && typeof a[key] === 'object'
              ? merge(a[key], b[key])
              : b[key],
        }),
      {}
    )
  )

export const compose = (...funcs) => {
  const fn = props =>
    funcs
      .map(fn => fn(props))
      .filter(Boolean)
      .reduce(merge, {})

  fn.propTypes = funcs.map(fn => fn.propTypes).reduce(merge, {})
  return fn
}

export const createMediaQuery = n => `@media screen and (min-width: ${px(n)})`

const getStyles = ({ props, style, value }) => {
  if (!isObject(value)) {
    return style(value)
  }

  // how to hoist this up??
  const breakpoints = get(props.theme, 'breakpoints') || defaultBreakpoints

  if (Array.isArray(value)) {
    const styles = style(value[0]) || {}
    for (let i = 1; i < value.length; i++) {
      const rule = style(value[i])
      if (rule) {
        const media = createMediaQuery(breakpoints[i - 1])
        styles[media] = rule
      }
    }
    return styles
  }

  const styles = {}
  for (let breakpoint in value) {
    const minWidth = breakpoints[breakpoint]
    if (!minWidth) {
      Object.assign(styles, style(value[breakpoint]))
    } else {
      const rule = style(value[breakpoint])
      const media = createMediaQuery(minWidth)
      styles[media] = rule
    }
  }

  return styles
}

export const style = ({
  prop,
  cssProperty,
  key,
  getter,
  transformValue,
  scale: defaultScale = {},
}) => {
  const css = cssProperty || prop
  const transform = transformValue || getter || noop
  const fn = props => {
    const value = props[prop]
    if (!is(value)) return null

    const scale = get(props.theme, key) || defaultScale
    const style = n =>
      is(n)
        ? {
            [css]: transform(get(scale, n) || n),
          }
        : null

    return getStyles({ props, style, value })
  }

  fn.propTypes = { [prop]: cloneFunc(propTypes.responsive) }

  return fn
}

export const util = {
  propTypes,
  defaultBreakpoints,
  is,
  num,
  px,
  get,
  themeGet,
  cloneFunc,
  merge,
  compose,
  createMediaQuery,
  style,
}

// space
//
// TODO
// - rewrite based on core style function
const isNegative = n => n < 0
const REG = /^[mp][trblxy]?$/
const properties = {
  m: 'margin',
  p: 'padding',
}
const directions = {
  t: 'Top',
  r: 'Right',
  b: 'Bottom',
  l: 'Left',
  x: ['Left', 'Right'],
  y: ['Top', 'Bottom'],
}

const getProperties = key => {
  const [a, b] = key.split('')
  const property = properties[a]
  const direction = directions[b] || ''
  return Array.isArray(direction)
    ? direction.map(dir => property + dir)
    : [property + direction]
}

const getValue = scale => n => {
  if (!num(n)) {
    return px(get(scale, n) || n)
  }
  const abs = Math.abs(n)
  const neg = isNegative(n)
  const value = scale[abs] || abs
  if (!num(value)) {
    return neg ? '-' + value : value
  }
  return px(value * (neg ? -1 : 1))
}

const defaultScale = [0, 4, 8, 16, 32, 64, 128, 256, 512]

export const space = props => {
  const keys = Object.keys(props)
    .filter(key => REG.test(key))
    .sort()
  const scale = get(props.theme, 'space') || defaultScale
  const getStyle = getValue(scale)

  return keys
    .map(key => {
      const value = props[key]
      const properties = getProperties(key)

      const style = n =>
        is(n)
          ? properties.reduce(
              (a, prop) => ({
                ...a,
                [prop]: getStyle(n),
              }),
              {}
            )
          : null

      return getStyles({ props, style, value })
    })
    .reduce(merge, {})
}

space.propTypes = {
  m: cloneFunc(propTypes.responsive),
  mt: cloneFunc(propTypes.responsive),
  mr: cloneFunc(propTypes.responsive),
  mb: cloneFunc(propTypes.responsive),
  ml: cloneFunc(propTypes.responsive),
  mx: cloneFunc(propTypes.responsive),
  my: cloneFunc(propTypes.responsive),
  p: cloneFunc(propTypes.responsive),
  pt: cloneFunc(propTypes.responsive),
  pr: cloneFunc(propTypes.responsive),
  pb: cloneFunc(propTypes.responsive),
  pl: cloneFunc(propTypes.responsive),
  px: cloneFunc(propTypes.responsive),
  py: cloneFunc(propTypes.responsive),
}


